### ISSUE TITLE
Add ManageIQ addon for quick dev setup for working with MIQ

### ISSUE THREAD

**Issue by NickLaMuro** [minishift/minishift#710]
Background
----------
The current developer setup process for adding the ManageIQ Openshift provider requires using Ansible and running on a external host (or a collection of VMs).  For minor testing purposes and simple end-to-end testing, this is a bit overkill, and includes a lot of external dependencies to get working.

Proposal
--------
Add an addon which configures openshift in a similar manner as the [Ansible playbook](https://github.com/openshift/openshift-ansible/blob/master/roles/openshift_manageiq/tasks/main.yaml) for manageiq (which is effectively just a collection of `oc` commands).  The resulting addon file should looks something like this:

```bash
# Name: manageiq
# Description:  Basic Users, Roles, and Policies necessary to have ManageIQ connect to Openshift

oc adm new-project 'management-infra' --description='Management Infrastructure' 

oc create serviceaccount 'management-admin' -n 'management-infra'
oc create serviceaccount 'inspector-admin' -n 'management-infra'

oc create -f 'management-infra-admin-cluster-role.json'
oc create -f 'hawkular-metrics-admin-cluster-role.json'

oc policy add-role-to-user -n management-infra admin -z management-admin
oc policy add-role-to-user -n management-infra management-infra-admin -z management-admin
oc adm policy add-cluster-role-to-user cluster-reader system:serviceaccount:management-infra:management-admin
oc adm policy add-scc-to-user privileged system:serviceaccount:management-infra:management-admin
oc adm policy add-cluster-role-to-user system:image-puller system:serviceaccount:management-infra:inspector-admin
oc adm policy add-scc-to-user privileged system:serviceaccount:management-infra:inspector-admin
oc adm policy add-cluster-role-to-user self-provisioner system:serviceaccount:management-infra:management-admin
oc adm policy add-cluster-role-to-user hawkular-metrics-admin system:serviceaccount:management-infra:management-admin
```

(there would also need to be two json files for creating the cluster roles, but I have omitted those here to save space)

Issues
------
There are currently a few issues surrounding the above that I have found in testing the above addon (which I realize is a fairly recent addition to the project):

1. It fails on the first line with the following error (note, this works fine when you run the `oc` command from the host):
  
  ```
  Applying addon manageiq:
  -- adm new-project 'management-infra' --description='Management Infrastructure'
  error: you must specify one argument: project name
  See 'oc adm new-project -h' for help and examples.
  E0406 17:44:18.967794   54419 start.go:232] Error executing addon commands Error executing command oc adm new-project 'management-infra' --description='Management Infrastructure
  ```
  
2. While not directly related, the current version of the Ansible playbooks have a pretty limited [requirement of version numbers](https://github.com/openshift/openshift-ansible/blob/master/roles/openshift_manageiq/tasks/main.yaml#L2-L4), and noticed in a [recent issue](https://github.com/minishift/minishift/issues/667), the versions of openshift that the the `rc` version of `minishift` now supports is essentially limited to the newest release candidate.  Now, this might not be a huge issue, and the only issue I have run into so far with this is changing of commands (`oc policy add-cluster-role-to-user` to `oc adm policy add-cluster-role-to-user` for example) which would break the playbooks in there current form, but not necessarily the API interaction per say.  More testing would be needed, but something to be aware of at a minimum.

I can put together the PR for adding the addon to the core group of addons if we decide that is necessary, but currently am blocked from being able to test if it is functioning as expected because of failing first command.

**Comment by jorgemoralespou** [comment]
@NickLaMuro Hi, you can use [oc-cluster](https://github.com/openshift-evangelists/oc-cluster-wrapper) as field for experimentation with different oc versions, as it supports all of them, and this way you can develop the addon, by looking into inconsistencies between versions.

Then it will be easier and faster to create the PR here, although not sure if this should be a default addon, as then for CDK this should be removed and/or replaced with a CFME one.

Out of interest, there's already a addon being worked (in oc-cluster is called plugin) but are basically the same, [here](https://github.com/openshift-evangelists/oc-cluster-wrapper/pull/59/files), so you can get some ideas as well.

**Comment by NickLaMuro** [comment]
@jorgemoralespou Thanks for the response.

I think my use case for the addon is slightly different then what might be designing the plugin for over in `oc-cluster`.  From what I have read/interpreted in that PR, you are setting up a project within openshift to actually run/manage CFME/MIQ instance within openshift, where my intent with this is to actually just give permissions openshift to allow ManageIQ to manage it that is already being run/developed on the dev's machine.

That said, to do it, I was reverse engineering the ansible playbook (linked in the description), and so I might have added more than was necessary to do it (for example, seems weird that I am adding a arbitrary project to the cluster just to allow it to make API calls against the cluster for inventory purposes).  I fully admit to being a bit of a novice with Ansible (come from a Chef background previously), and am only just getting my feet wet with openshift, so take that into consideration here as well.

I will try and see if some of this can be removed from the addon as it does seem a little excessive, and in doing so, that might remove the first issue from above from being relevant.  That said, still seems weird that a project can't be created from within an `addon` still, but maybe I am missing something.

**Comment by NickLaMuro** [comment]
> That said, still seems weird that a project can't be created from within an `addon` still, but maybe I am missing something.

After doing a bit more reading, looks like the project is used as a namespace/grouping mechanic that is baked into the mechanics of `servicesaccounts` in openshift, so you can ignore that previous statement 😄

**Comment by NickLaMuro** [comment]
After trying a few things out, I realized that the `openshift-infra` project already existed in the deployment of minishift.  After updating the above script to swap out uses of the `management-infra` namespace to `openshift-infra` and removing the first line to create the new project, the modified addon was able to complete successfully, and allowed communication between openshift and manageiq to function properly.

After a bit of trial and error (and trying to get a more isolated example for you both to confirm this wasn't just an issue for me), I think I have isolated the cause of the bug with creating a project, and it is the whitespace in the description for the command.  Take the following two addons as an example:

**Works:**
```console
$ mkdir -p ./tmp/minishift_addons/foo
$ echo -e # Name: foo\n# Description: create foo project\n\noc adm new-project foo --description='Bar' > ./tmp/minishift_addons/foo/foo.addon
$ minishift addons install ./tmp/minishift_addons/foo
$ minishift addons enable foo
$ minishift start
...
Applying addon foo:
-- adm new-project foo --description='Bar'
Created project foo
  [306.720335ms]
```

**Fails**
```console
$ mkdir -p ./tmp/minishift_addons/foo
$ echo -e # Name: foo\n# Description: create foo project\n\noc adm new-project foo --description='Bar Baz' > ./tmp/minishift_addons/foo/foo.addon
$ minishift addons install ./tmp/minishift_addons/foo
$ minishift addons enable foo
$ minishift start
...
Applying addon foo:
-- adm new-project foo --description='Bar Baz'
error: you must specify one argument: project name
See 'oc adm new-project -h' for help and examples.
E0407 17:09:06.185670   20557 start.go:232] Error executing addon commands Error executing command oc adm new-project foo --description='Bar Baz'.
```

My guess is that the command parsing that was added with the [addon feature](https://github.com/minishift/minishift/pull/625) was implemented in such a way that it is splitting that into two args, instead of one, causing a generic error message to be returned from the `oc` command (this is a guess... haven't fully grok'd the code from #625 ).  I will open up a separate issue for this since I think this is out of scope of the original feature request, and I think I have sufficiently been able to prove it wasn't user error (whew!).

I think I can also update the `oc adm new-project` command in the description to just pass a non-whitespace description for now as a workaround.

**Comment by jorgemoralespou** [comment]
@NickLaMuro. Thanks for the analysis, and for catching that bug. 

With respect to this request I'm wondering if just enabling minishift to be managed by ManageIQ is a very common use case to make it part of the default addons. 

One of the goals of the addons will be to allow one to install externally provided addons, so you can just provide a command to your users like:

minishift addon install <git location of my addon>

**Comment by NickLaMuro** [comment]
> With respect to this request I'm wondering if just enabling minishift to be managed by ManageIQ is a very common use case to make it part of the default addons.

@jorgemoralespou This is one question I can't give you a solid answer to :smile:.  I can say that the use case does exist from a production perspective, but how common it is I frankly can't give you the numbers on that since I just don't know.

That said, I could turn this into a gist just fine, and that might be the more pragmatic way of doing this for the time being.  You may have noticed, but I am currently also working on some documentation over in the ManageIQ/guides for getting this in place for a developer there as well:

https://github.com/ManageIQ/guides/issues/204

And the intent of this request was to be able to simplify the instructions a bit, as well as instill a bit of confidence from the user that they weren't just "installing some random script from the internet".  That said, if a git url is an acceptable argument for that command (I didn't realize that myself, though, looking over the code... it might not be that simple...), that might solve our use case just fine, since that is barely a difference in command length.  Even requiring a step to clone the gist, or working it into an existing step, shouldn't be that big of a deal.

**Comment by jorgemoralespou** [comment]
@NickLaMuro Using git url as source is not yet a possibility (AFAIK) but will soon be. Am I correct @hferentschik?

Also, I would think we need to think how we will be asking people to contribute addons as not all can go into the default addons, which will mean they will go through the same lifecycle of minishift, which can be a burden. Just think on:

what happens if there is a bug in one addon? should minishift make a new release? should you make the addon users wait for a new release?

I would think that there is 2 kind of addons, one the core, which are shipped with the product, and then the additional, which can be provided by any user at any time (decoupling their lifecycle with that of minishift). I would think that this one should be of the second type.

Then, it's up to the minishift team to have a project aggregation of these additional addons (in a semi official way), something like minishift/additional-addons in a github repo, where they aggregate content coming from multiple different sources, or that you host the addon with your code (or application), not in a gist, but in a full source repo in your github. e.g. Manageiq/minishift-addon. This way it's more official than a script provided in a documentation and you can have full lifecycle plus issue tracking.

**Comment by hferentschik** [comment]
> My guess is that the command parsing that was added with the addon feature was implemented in such a way that it is splitting that into two args, instead of one, causing a generic error message to be returned from the oc command. I will open up a separate issue

+1, sounds like a reasonable explanation. Please open a dedicated issue and we will investigate. The parser will actually take the line as is, so I am a bit surprised. We might need some additional escaping magic internally.

**Comment by hferentschik** [comment]
@NickLaMuro, I absolutely love the you picked up the add-on mechanism and that in the end you got it to work. Regarding adding it to the default addon list, as @jorgemoralespou is saying, we need to have a discussion on how to do this.

We want for sure more default addons, but I feel with default add-ons there is a certain expectation on the Minishift team to ensure that all is working. I am not sure whether we have the bandwidth atm to make this happen.

I think the first step forward is to implement to install add-ons by pointing to a URL. I think this would go a long way. Right now you can only install by pointing to a directory on the filesystem which is limiting. 

Further down the track one could think about setting up a repository of sort, basically some sort of repo of addons which live outside the main Minishift life cycle and for which the Minishift team also don't give the same guarantees in terms of stability.

WDYT?

**Comment by NickLaMuro** [comment]
@jorgemoralespou @hferentschik Sorry for my delay in response, but I agree with the general consensus that this probably doesn't make sense to live here as a default.  Not only does it make it hard to fix/debug, but it also adds extra weight to the binary as it is something additional that needs to be added to the project.  I think making this an external resource in conjunction with a mechanism allows remote addons to be installed is a much better long term goal then adding this as a default plugin.

> Please open a dedicated issue and we will investigate. The parser will actually take the line as is, so I am a bit surprised. We might need some additional escaping magic internally.

@hferentschik Did a bit of spelunking myself and I think I have isolated where the failure is.  Still working on getting some stuff together to confirm a working prototype for you to reduce the legwork necessary to get the fix in (which I will let you decide on how useful/worthless it is).  Hopefully the sample will provide additional corner case prevention and reduce future bugs.

Will have something up later today hopefully regarding that bug.

**Comment by NickLaMuro** [comment]
Bug issue has been opened up here:  https://github.com/minishift/minishift/issues/729

**Comment by praveenkumar** [comment]
@NickLaMuro I hope this is fix for you can we close this issue now?

**Comment by NickLaMuro** [comment]
@praveenkumar Must have missed [this](https://github.com/minishift/minishift/issues/710#issuecomment-305165716), but I think that https://github.com/minishift/minishift/issues/635 will be solving this need.

#729 was just solving a bug with the addon code I had provided above, but not solving the core of this issue.

But again, because of #635, and some of the good points about having to maintain this within the release cycle of the binary, I think that this can be closed in favor of #635 .

**Comment by budhrg** [comment]
Thanks @NickLaMuro  for update. 

I am closing it in favor of #635  as per your comment.
